File: android/src/main/java/com/easemob/im_flutter_sdk/ContactManagerWrapper.java
package com.easemob.im_flutter_sdk;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.hyphenate.EMContactListener;
import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMContact;
import com.hyphenate.chat.EMCursorResult;
import com.hyphenate.exceptions.HyphenateException;

import org.json.JSONException;
import org.json.JSONObject;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

public class ContactManagerWrapper extends Wrapper implements MethodCallHandler {

    private EMContactListener contactListener;

    ContactManagerWrapper(FlutterPlugin.FlutterPluginBinding flutterPluginBinding, String channelName) {
        super(flutterPluginBinding, channelName);
        registerEaseListener();
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {

        JSONObject param = (JSONObject) call.arguments;
        try {
            if (MethodKey.addContact.equals(call.method)) {
                addContact(param, call.method, result);
            } else if (MethodKey.deleteContact.equals(call.method)) {
                deleteContact(param, call.method, result);
            } else if (MethodKey.getAllContactsFromServer.equals(call.method)) {
                getAllContactsFromServer(param, call.method, result);
            } else if (MethodKey.getAllContactsFromDB.equals(call.method)) {
                getAllContactsFromDB(param, call.method, result);
            } else if (MethodKey.addUserToBlockList.equals(call.method)) {
                addUserToBlockList(param, call.method, result);
            } else if (MethodKey.removeUserFromBlockList.equals(call.method)) {
                removeUserFromBlockList(param, call.method, result);
            } else if (MethodKey.getBlockListFromServer.equals(call.method)) {
                getBlockListFromServer(param, call.method, result);
            } else if (MethodKey.getBlockListFromDB.equals(call.method)) {
                getBlockListFromDB(param, call.method, result);
            } else if (MethodKey.acceptInvitation.equals(call.method)) {
                acceptInvitation(param, call.method, result);
            } else if (MethodKey.declineInvitation.equals(call.method)) {
                declineInvitation(param, call.method, result);
            } else if (MethodKey.getSelfIdsOnOtherPlatform.equals(call.method)) {
                getSelfIdsOnOtherPlatform(param, call.method, result);
            }else if (MethodKey.getAllContacts.equals(call.method)) {
                getAllContacts(param, call.method, result);
            }else if (MethodKey.setContactRemark.equals(call.method)) {
                setContactRemark(param, call.method, result);
            }else if (MethodKey.getContact.equals(call.method)) {
                getContact(param, call.method, result);
            }else if (MethodKey.fetchAllContacts.equals(call.method)) {
                fetchAllContacts(param, call.method, result);
            }else if (MethodKey.fetchContacts.equals(call.method)) {
                fetchContacts(param, call.method, result);
            }
            else {
                super.onMethodCall(call, result);
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private void addContact(JSONObject param, String channelName, Result result) throws JSONException {
        String username = param.getString("username");
        String reason = null;
        if(param.has("reason")) {
            reason = param.getString("reason");
        }
        String finalReason = reason;
        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().addContact(username, finalReason);
                onSuccess(result, channelName, username);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void deleteContact(JSONObject param, String channelName, Result result) throws JSONException {
        String username = param.getString("username");
        boolean keepConversation = param.getBoolean("keepConversation");
        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().deleteContact(username, keepConversation);
                onSuccess(result, channelName, username);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void getAllContactsFromServer(JSONObject params, String channelName, Result result) throws JSONException {
        asyncRunnable(() -> {
            try {
                List contacts = EMClient.getInstance().contactManager().getAllContactsFromServer();
                onSuccess(result, channelName, contacts);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void getAllContactsFromDB(JSONObject params, String channelName, Result result) throws JSONException {
        asyncRunnable(() -> {
            try {
                List contacts = EMClient.getInstance().contactManager().getContactsFromLocal();
                onSuccess(result, channelName, contacts);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void addUserToBlockList(JSONObject params, String channelName, Result result) throws JSONException {
        String username = params.getString("username");
        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().addUserToBlackList(username, false);
                onSuccess(result, channelName, username);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void removeUserFromBlockList(JSONObject params, String channelName, Result result) throws JSONException {
        String username = params.getString("username");
        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().removeUserFromBlackList(username);
                onSuccess(result, channelName, username);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void getBlockListFromServer(JSONObject params, String channelName, Result result) throws JSONException {
        asyncRunnable(() -> {
            try {
                List contacts = EMClient.getInstance().contactManager().getBlackListFromServer();
                onSuccess(result, channelName, contacts);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void getBlockListFromDB(JSONObject params, String channelName, Result result) throws JSONException {
        asyncRunnable(()->{
            List contacts = EMClient.getInstance().contactManager().getBlackListUsernames();
            onSuccess(result, channelName, contacts);
        });
    }

    private void acceptInvitation(JSONObject params, String channelName, Result result) throws JSONException {
        String username = params.getString("username");
        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().acceptInvitation(username);
                onSuccess(result, channelName, username);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void declineInvitation(JSONObject params, String channelName, Result result) throws JSONException {
        String username = params.getString("username");
        asyncRunnable(() -> {
            try {
                EMClient.getInstance().contactManager().declineInvitation(username);
                onSuccess(result, channelName, username);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void getSelfIdsOnOtherPlatform(JSONObject params, String channelName, Result result) throws JSONException {
        asyncRunnable(() -> {
            try {
                List platforms = EMClient.getInstance().contactManager().getSelfIdsOnOtherPlatform();
                onSuccess(result, channelName, platforms);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void getAllContacts(JSONObject params, String channelName, Result result) throws JSONException {
        EMClient.getInstance().contactManager().asyncFetchAllContactsFromLocal(new EMValueWrapperCallBack<List<EMContact>>(result, channelName) {
            @Override
            public void onSuccess(List<EMContact> list){
                List<Map> contactList = new ArrayList<>();
                for (EMContact contact : list) {
                    contactList.add(ContactHelper.toJson(contact));
                }
                updateObject(contactList);
            }
        });
    }

    private void setContactRemark(JSONObject params, String channelName, Result result) throws JSONException {
        String userId = params.optString("userId");
        String remark = params.optString("remark");
        EMClient.getInstance().contactManager().asyncSetContactRemark(userId, remark, new EMWrapperCallBack(result, channelName, null));
    }

    private void getContact(JSONObject params, String channelName, Result result) throws JSONException {
        String userId = params.optString("userId");
        asyncRunnable(() -> {
            try {
                EMContact contact = EMClient.getInstance().contactManager().fetchContactFromLocal(userId);
                if (contact != null) {
                    onSuccess(result, channelName, ContactHelper.toJson(contact));
                }else {
                    onSuccess(result, channelName, null);
                }
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void fetchAllContacts(JSONObject params, String channelName, Result result) throws JSONException {
        EMClient.getInstance().contactManager().asyncFetchAllContactsFromServer(new EMValueWrapperCallBack<List<EMContact>>(result, channelName) {
            @Override
            public void onSuccess(List<EMContact> list){
                List<Map> contactList = new ArrayList<>();
                for (EMContact contact : list) {
                    contactList.add(ContactHelper.toJson(contact));
                }
                updateObject(contactList);
            }
        });
    }

    private void fetchContacts(JSONObject params, String channelName, Result result) throws JSONException {
        int pageSize = params.getInt("pageSize");
        String cursor = null;
        if (params.has("cursor")) {
            cursor = params.getString("cursor");
        }

        EMClient.getInstance().contactManager().asyncFetchAllContactsFromServer(pageSize, cursor, new EMValueWrapperCallBack<EMCursorResult<EMContact>>(result, channelName){
            @Override
            public void onSuccess(EMCursorResult<EMContact> object) {
                super.updateObject(CursorResultHelper.toJson(object));
            }
        });
    }


    private void registerEaseListener() {

        if (contactListener != null) {
            EMClient.getInstance().contactManager().removeContactListener(contactListener);
        }

        contactListener = new EMContactListener() {
            @Override
            public void onContactAdded(String userName) {
                ListenerHandle.getInstance().addHandle(
                        ()-> {
                            Map<String, Object> data = new HashMap<>();
                            data.put("type", "onContactAdded");
                            data.put("username", userName);
                            post(() -> channel.invokeMethod(MethodKey.onContactChanged, data));
                        }
                );
            }

            @Override
            public void onContactDeleted(String userName) {
                ListenerHandle.getInstance().addHandle(
                        ()-> {
                            Map<String, Object> data = new HashMap<>();
                            data.put("type", "onContactDeleted");
                            data.put("username", userName);
                            post(() -> channel.invokeMethod(MethodKey.onContactChanged, data));
                        }
                );
            }

            @Override
            public void onContactInvited(String userName, String reason) {
                ListenerHandle.getInstance().addHandle(
                        ()-> {
                            Map<String, Object> data = new HashMap<>();
                            data.put("type", "onContactInvited");
                            data.put("username", userName);
                            data.put("reason", reason);
                            post(() -> channel.invokeMethod(MethodKey.onContactChanged, data));
                        }
                );
            }

            @Override
            public void onFriendRequestAccepted(String userName) {
                ListenerHandle.getInstance().addHandle(
                        ()-> {
                            Map<String, Object> data = new HashMap<>();
                            data.put("type", "onFriendRequestAccepted");
                            data.put("username", userName);
                            post(() -> channel.invokeMethod(MethodKey.onContactChanged, data));
                        }
                );
            }

            @Override
            public void onFriendRequestDeclined(String userName) {
                ListenerHandle.getInstance().addHandle(
                        ()-> {
                            Map<String, Object> data = new HashMap<>();
                            data.put("type", "onFriendRequestDeclined");
                            data.put("username", userName);
                            post(() -> channel.invokeMethod(MethodKey.onContactChanged, data));
                        }
                );
            }
        };

        EMClient.getInstance().contactManager().setContactListener(contactListener);
    }

    @Override
    public void unRegisterEaseListener() {
        EMClient.getInstance().contactManager().removeContactListener(contactListener);
    }
}

File: ios/Classes/ContactManagerWrapper.m
//
//  EMContactManagerWrapper.m
//
//
//  Created by 杜洁鹏 on 2019/10/8.
//

#import "ContactManagerWrapper.h"
#import "MethodKeys.h"
#import "ListenerHandle.h"
#import "ContactHelper.h"
#import "CursorResultHelper.h"

@interface ContactManagerWrapper () <EMContactManagerDelegate>

@end

@implementation ContactManagerWrapper
- (instancetype)initWithChannelName:(NSString *)aChannelName
                          registrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    if(self = [super initWithChannelName:aChannelName
                               registrar:registrar]) {
        [EMClient.sharedClient.contactManager addDelegate:self delegateQueue:nil];
    }
    return self;
}

- (void)unRegisterEaseListener {
    [EMClient.sharedClient.contactManager removeDelegate:self];
}


#pragma mark - FlutterPlugin

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    
    if ([ChatAddContact isEqualToString:call.method]) {
        [self addContact:call.arguments
             channelName:call.method
                  result:result];
    } else if ([ChatDeleteContact isEqualToString:call.method]) {
        [self deleteContact:call.arguments
                channelName:call.method
                     result:result];
    } else if ([ChatGetAllContactsFromServer isEqualToString:call.method]) {
        [self getAllContactsFromServer:call.arguments
                           channelName:call.method
                                result:result];
    } else if ([ChatGetAllContactsFromDB isEqualToString:call.method]) {
        [self getAllContactsFromDB:call.arguments
                       channelName:call.method
                            result:result];
    } else if ([ChatAddUserToBlockList isEqualToString:call.method]) {
        [self addUserToBlockList:call.arguments
                     channelName:call.method
                          result:result];
    } else if ([ChatRemoveUserFromBlockList isEqualToString:call.method]) {
        [self removeUserFromBlockList:call.arguments
                          channelName:call.method
                               result:result];
    } else if ([ChatGetBlockListFromServer isEqualToString:call.method]) {
        [self getBlockListFromServer:call.arguments
                         channelName:call.method
                              result:result];
    } else if ([ChatGetBlockListFromDB isEqualToString:call.method]){
        [self getBlockListFromDB:call.arguments
                     channelName:call.method
                          result:result];
    } else if ([ChatAcceptInvitation isEqualToString:call.method]) {
        [self acceptInvitation:call.arguments
                   channelName:call.method
                        result:result];
    } else if ([ChatDeclineInvitation isEqualToString:call.method]) {
        [self declineInvitation:call.arguments
                    channelName:call.method
                         result:result];
    } else if ([ChatGetSelfIdsOnOtherPlatform isEqualToString:call.method]) {
        [self getSelfIdsOnOtherPlatform:call.arguments
                            channelName:call.method
                                 result:result];
    } else if ([ChatGetAllContacts isEqualToString:call.method]) {
        [self getAllContacts:call.arguments
                            channelName:call.method
                                 result:result];
    } else if ([ChatSetContactRemark isEqualToString:call.method]) {
        [self setContactRemark:call.arguments
                            channelName:call.method
                                 result:result];
    } else if ([ChatGetContact isEqualToString:call.method]) {
        [self getContact:call.arguments
                            channelName:call.method
                                 result:result];
    } else if ([ChatFetchAllContacts isEqualToString:call.method]) {
        [self fetchAllContacts:call.arguments
                            channelName:call.method
                                 result:result];
    } else if ([ChatFetchContacts isEqualToString:call.method]) {
        [self fetchContacts:call.arguments
                            channelName:call.method
                                 result:result];
    } else {
        [super handleMethodCall:call result:result];
    }
}


#pragma mark - Actions
- (void)addContact:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    NSString *reason = param[@"reason"];
    [EMClient.sharedClient.contactManager addContact:username
                                             message:reason
                                          completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
}

- (void)deleteContact:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    BOOL keepConversation = [param[@"keepConversation"] boolValue];
    [EMClient.sharedClient.contactManager deleteContact:username
                                   isDeleteConversation:keepConversation
                                             completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
}

- (void)getAllContactsFromServer:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    [EMClient.sharedClient.contactManager getContactsFromServerWithCompletion:^(NSArray *aList, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aList];
    }];
}

- (void)getAllContactsFromDB:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSArray *aList = EMClient.sharedClient.contactManager.getContacts;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:aList];
}


- (void)addUserToBlockList:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    [EMClient.sharedClient.contactManager addUserToBlackList:username
                                                  completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
}

- (void)removeUserFromBlockList:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    [EMClient.sharedClient.contactManager removeUserFromBlackList:username
                                                       completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
    
}

- (void)getBlockListFromServer:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    [EMClient.sharedClient.contactManager getBlackListFromServerWithCompletion:^(NSArray *aList, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aList];
    }];
}

- (void)getBlockListFromDB:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    
    NSArray * list = [EMClient.sharedClient.contactManager getBlackList];
    [self wrapperCallBack:result channelName:aChannelName error:nil object:list];
}

- (void)acceptInvitation:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    [EMClient.sharedClient.contactManager approveFriendRequestFromUser:username
                                                            completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
}

- (void)declineInvitation:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    NSString *username = param[@"username"];
    [EMClient.sharedClient.contactManager declineFriendRequestFromUser:username
                                                            completion:^(NSString *aUsername, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aUsername];
    }];
}

- (void)getSelfIdsOnOtherPlatform:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    [EMClient.sharedClient.contactManager getSelfIdsOnOtherPlatformWithCompletion:^(NSArray *aList, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aList];
    }];
}


- (void)getAllContacts:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    
    NSArray* contacts = [EMClient.sharedClient.contactManager getAllContacts];
    
    NSMutableArray *contactList = [NSMutableArray array];
    for (EMContact *contact in contacts) {
        [contactList addObject:[contact toJson]];
    }
    
    [self wrapperCallBack:result
              channelName:aChannelName
                    error:nil
                   object:contactList];
}

///
- (void)setContactRemark:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    
    NSString *userId = param[@"userId"];
    NSString *remark = param[@"remark"];
    
    [EMClient.sharedClient.contactManager setContactRemark:userId
                                                    remark:remark
                                                completion:^(EMContact * _Nullable contact, EMError * _Nullable aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}


- (void)getContact:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    NSString *userId = param[@"userId"];
    EMContact *contact = [EMClient.sharedClient.contactManager getContact:userId];
    [self wrapperCallBack:result
              channelName:aChannelName
                    error:nil
                   object:[contact toJson]];
}

- (void)fetchAllContacts:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self)weakSelf = self;
    [EMClient.sharedClient.contactManager getAllContactsFromServerWithCompletion:^(NSArray<EMContact *> * _Nullable aList, EMError * _Nullable aError) {
        NSMutableArray *contactList = [NSMutableArray array];
        for (EMContact *contact in aList) {
            [contactList addObject:[contact toJson]];
        }
        
        [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:aError
                       object:contactList];
    }];
}


- (void)fetchContacts:(NSDictionary *)param channelName:(NSString *)aChannelName result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *cursor = param[@"cursor"];
    int pageSize = [param[@"pageSize"] intValue];
    [EMClient.sharedClient.contactManager getContactsFromServerWithCursor:cursor
                                                                 pageSize:pageSize
                                                               completion:^(EMCursorResult<EMContact *> * _Nullable aResult, EMError * _Nullable aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:[aResult toJson]];
    }];
}


#pragma mark - EMContactManagerDelegate

- (void)friendshipDidAddByUser:(NSString *)aUsername {
    __weak typeof(self) weakSelf = self;
    [ListenerHandle.sharedInstance addHandle:^{
        NSDictionary *map = @{
            @"type":@"onContactAdded",
            @"username":aUsername
        };
        [weakSelf.channel invokeMethod:ChatOnContactChanged arguments:map];
    }];

}

- (void)friendshipDidRemoveByUser:(NSString *)aUsername {
    __weak typeof(self) weakSelf = self;
    [ListenerHandle.sharedInstance addHandle:^{
        NSDictionary *map = @{
            @"type":@"onContactDeleted",
            @"username":aUsername
        };
        [weakSelf.channel invokeMethod:ChatOnContactChanged arguments:map];
    }];
}

- (void)friendRequestDidReceiveFromUser:(NSString *)aUsername
                                message:(NSString *)aMessage {
    __weak typeof(self) weakSelf = self;
    [ListenerHandle.sharedInstance addHandle:^{
        NSDictionary *map = @{
            @"type":@"onContactInvited",
            @"username":aUsername,
            @"reason":aMessage
        };
        [weakSelf.channel invokeMethod:ChatOnContactChanged arguments:map];
    }];

}

- (void)friendRequestDidApproveByUser:(NSString *)aUsername {
    __weak typeof(self) weakSelf = self;
    [ListenerHandle.sharedInstance addHandle:^{
        NSDictionary *map = @{
            @"type":@"onFriendRequestAccepted",
            @"username":aUsername
        };
        [weakSelf.channel invokeMethod:ChatOnContactChanged arguments:map];
    }];
}

- (void)friendRequestDidDeclineByUser:(NSString *)aUsername {
    __weak typeof(self) weakSelf = self;
    [ListenerHandle.sharedInstance addHandle:^{
        NSDictionary *map = @{
            @"type":@"onFriendRequestDeclined",
            @"username":aUsername
        };
        [weakSelf.channel invokeMethod:ChatOnContactChanged arguments:map];
    }];
}

@end
